### ISSUE TITLE
Crash in iOS 12.4+ Starting with Kingfisher 6.2.0

### ISSUE THREAD

**Issue by bb-git** [onevcat/Kingfisher#1717]
### Issue Description
The App Crashes directly at launch. There seems to be a connection with RealmSwift v10. But I didn't find out, what it is.

#### What

```
dyld: lazy symbol binding failed: can't resolve symbol _$s7Combine9PublishedVMa in <redacted>/Frameworks/Kingfisher.framework/Kingfisher because dependent dylib #3 could not be loaded
dyld: can't resolve symbol _$s7Combine9PublishedVMa in<redacted>/Frameworks/Kingfisher.framework/Kingfisher because dependent dylib #3 could not be loaded
```

I will try to reproduce it in a sample project

#### Other Comment
Going back to 6.1.1 the App starts without any issues.

**Comment by onevcat** [comment]
Not sure how this happens. But it seems to be related to Combine? Maybe you can try to link Combine explicitly to see if it can be solved.

**Comment by bb-git** [comment]
I linked it in the build phases, but it still crashes. It's necessary to create a Realm instance `let realm = try! Realm()` and it crashes.

**Comment by tackgyu** [comment]
https://stackoverflow.com/questions/63190381/how-to-conditionally-use-combine-published-when-minimum-deployment-target-is-se

Run time error occurs on our proj too (except only in unit test bundle though) after https://github.com/onevcat/Kingfisher/pull/1642 merged from where `Combine.Published` property wrapper is introduced to `Kingfisher`. We suspect the property wrapper `Published` in `Combine` framework actually does not work on iOS 12 or below.

We dropped investigating this case as it does not affect production but I will leave this comment for someone undergoing the same issue.

**Comment by tackgyu** [comment]
FYI. tested on Xcode 12.2, Swift 5.3.1, Kingfisher built via Carthage

**Comment by alxterekh** [comment]
Has the same issue on Xcode 12.5  and iOS 12.4

**Comment by onevcat** [comment]
Thanks all the reports and investigation.

I myself cannot reproduce this crash, so I guess it is might be something on how the framework is built (Maybe Carthage? I didn't use it for a long time now so not sure can it be related.)

Following the answer from the SO post [mentioned above](https://stackoverflow.com/questions/63190381/how-to-conditionally-use-combine-published-when-minimum-deployment-target-is-se), I created a branch (`fix/runtime-combine-error`) in which the `@Published` eliminated.

I kindly ask if anyone can help to verify how that fix can work or not since I do not have a way to try myself. If it fixes the issue, I will prepare a patch release for it soon.

Thank you all!

**Comment by bb-git** [comment]
👍 Working for me. My Unit Test completed successfully. I could reproduce it with Cocoapods installing Realm 10.7.6 and Kingfisher 6.2+. Crahs happend on first init of the Database `let realm = try! Realm()`

**Comment by tikitu** [comment]
I can confirm 
* that 6.3.0 produces a runtime crash for me on iOS 12.4
* that the `fix/runtime-combine-error` branch fixes that crash

If you can make a patch release with those changes that would be very lovely indeed. (I think I see a `@ViewBuilder` sneaked in, in `KFImage`: it doesn't cause any problems for me, but you might want to double-check if it was intended.)

As for the root cause, it's not Realm (which we don't use). There appears to be a Swift/compiler bug that somehow lets types leak outside `@available` and friends, so it goes looking for type metadata which isn't available on that platform version. ([Another sufferer](https://stackoverflow.com/questions/62091996/swift-crash-in-swift-getsingletonmetadata-type-metadata-accessor).)

**Comment by onevcat** [comment]
@tikitu

Thanks for the feedback!

It is quite annoying that I can never reproduce it myself...So may I ask some questions in detail:

#### Does this issue only happen on a real device but not on simulator? 

I cannot find an iOS 12 right now, so if I want to know some more about the reproducing steps.

#### Do you have a chance to try the latest `7.0.0-beta.1` tag? 

There is almost no API changes for UIKit part in that major version, so it should be easy to give it a shot. (I suspect the same would happen since `@Published` is still used there.) If that does not change in the new version, I have to do that for v7 too, which is, I have to say, very dirty.

If we can confirm that the issue only happens on an iOS 12 real devices but not on simulator, I think I have to seek an old device in my cabinet... 😓

**Comment by tikitu** [comment]
Hi @onevcat, I'm getting this in a simulator. Xcode Version 12.5 (12E262), my project has a deployment target of iOS 12.3, simulator is iPhone 6 with iOS 12.4. I'm sorry I can't share the project with you, and unfortunately it's rather huge. I'll have a go on the 7.0.0 tag also and let you know.

**Comment by tikitu** [comment]
(Oh any by the way: we're using cocoapods for dependency management.)

**Comment by onevcat** [comment]
Thanks for the detail information. ❤️  

> I'm sorry I can't share the project with you

Sure, never mind! it is never easy or even possible. Waiting for the result and I will try it myself based on your setups!

**Comment by tikitu** [comment]
@onevcat for me the 7.0.0-beta.1 tag works without crash. I see the `ImageBinder` availability check changed to iOS 14, maybe it's only the iOS 13 `@Published` which is problematic? But now I'm really only guessing...

**Comment by tikitu** [comment]
The line of our code that triggers the crash is nothing more than putting a string/string key/value pair into a dictionary typed `[AnyHashable: Any]`. None of this has anything remotely to do with Kingfisher. That kicks off a call to `__swift_instantiateConcreteTypeFromMangledName` (I'm not sure why), which (several frames down) does a `swift_conformsToProtocol` check (on something, I can't read the assembly) which calls into the "type metadata accessor for KFImage.ImageBinder". Since you've guarded `ImageBinder` with `@available`, I guess this should not be happening. 😬

**Comment by tikitu** [comment]
The call to `TempImageServices.setUserAgent(_:)` at frame 24 is our code, and it's just a dictionary setter. After that it gets weird quick :-)

```Thread 1 Queue : com.apple.main-thread (serial)
#0	0x00000001124c1a8a in __abort_with_payload ()
#1	0x00000001124ea831 in abort_with_payload_wrapper_internal ()
#2	0x00000001124ea863 in abort_with_payload ()
#3	0x00000001083398c1 in abort_with_payload ()
#4	0x0000000108330fe0 in dyld::halt(char const*) ()
#5	0x000000010833110a in dyld::fastBindLazySymbol(ImageLoader**, unsigned long) ()
#6	0x000000011241f122 in dyld_stub_binder ()
#7	0x000000010c551000 in 0x10c551000 ()
#8	0x00000001241be213 in swift_getSingletonMetadata ()
#9	0x000000010c437658 in type metadata accessor for KFImage.ImageBinder ()
#10	0x00000001241d66dc in swift::TargetProtocolConformanceDescriptor<swift::InProcess>::getCanonicalTypeMetadata() const ()
#11	0x00000001241d77b4 in swift_conformsToSwiftProtocolImpl(swift::TargetMetadata<swift::InProcess> const*, swift::TargetProtocolDescriptor<swift::InProcess> const*, llvm::StringRef) ()
#12	0x00000001241d7598 in swift_conformsToProtocol ()
#13	0x00000001241abf7a in swift::_conformsToProtocol(swift::OpaqueValue const*, swift::TargetMetadata<swift::InProcess> const*, swift::TargetProtocolDescriptorRef<swift::InProcess>, swift::TargetWitnessTable<swift::InProcess> const**) ()
#14	0x00000001241d6e58 in swift::_checkGenericRequirements(llvm::ArrayRef<swift::TargetGenericRequirementDescriptor<swift::InProcess> >, std::__1::vector<void const*, std::__1::allocator<void const*> >&, std::__1::function<swift::TargetMetadata<swift::InProcess> const* (unsigned int, unsigned int)>, std::__1::function<swift::TargetWitnessTable<swift::InProcess> const* (swift::TargetMetadata<swift::InProcess> const*, unsigned int)>) ()
#15	0x00000001241d3743 in (anonymous namespace)::DecodedMetadataBuilder::createBoundGenericType(swift::TargetContextDescriptor<swift::InProcess> const*, llvm::ArrayRef<swift::TargetMetadata<swift::InProcess> const*>, swift::TargetMetadata<swift::InProcess> const*) const ()
#16	0x00000001241d2da0 in swift::Demangle::TypeDecoder<(anonymous namespace)::DecodedMetadataBuilder>::decodeMangledType(swift::Demangle::Node* const&) ()
#17	0x00000001241d20dc in swift::Demangle::TypeDecoder<(anonymous namespace)::DecodedMetadataBuilder>::decodeMangledType(swift::Demangle::Node* const&) ()
#18	0x00000001241d19de in swift_getTypeByMangledNodeImpl(swift::MetadataRequest, swift::Demangle::Demangler&, swift::Demangle::Node*, std::__1::function<swift::TargetMetadata<swift::InProcess> const* (unsigned int, unsigned int)>, std::__1::function<swift::TargetWitnessTable<swift::InProcess> const* (swift::TargetMetadata<swift::InProcess> const*, unsigned int)>) ()
#19	0x00000001241d1806 in swift::swift_getTypeByMangledNode(swift::MetadataRequest, swift::Demangle::Demangler&, swift::Demangle::Node*, std::__1::function<swift::TargetMetadata<swift::InProcess> const* (unsigned int, unsigned int)>, std::__1::function<swift::TargetWitnessTable<swift::InProcess> const* (swift::TargetMetadata<swift::InProcess> const*, unsigned int)>) ()
#20	0x00000001241d1e73 in swift_getTypeByMangledNameImpl(swift::MetadataRequest, llvm::StringRef, std::__1::function<swift::TargetMetadata<swift::InProcess> const* (unsigned int, unsigned int)>, std::__1::function<swift::TargetWitnessTable<swift::InProcess> const* (swift::TargetMetadata<swift::InProcess> const*, unsigned int)>) ()
#21	0x00000001241d0006 in swift::swift_getTypeByMangledName(swift::MetadataRequest, llvm::StringRef, std::__1::function<swift::TargetMetadata<swift::InProcess> const* (unsigned int, unsigned int)>, std::__1::function<swift::TargetWitnessTable<swift::InProcess> const* (swift::TargetMetadata<swift::InProcess> const*, unsigned int)>) ()
#22	0x00000001241d01a5 in swift_getTypeByMangledNameInContext ()
#23	0x00000001102682ad in __swift_instantiateConcreteTypeFromMangledName ()
#24	0x0000000110274e66 in TempImageServices.setUserAgent(_:) at /Users/tikitu/work/minddistrict/code/ios-minddistrict/Abstractions/Image downloads/Image downloads.services.swift:170
#25	0x0000000110275c7a in protocol witness for ImageServicesType.setUserAgent(_:) in conformance TempImageServices ()
#26	0x000000010fbd745f in GlobalAppState.init(network:storage:htmlFormatter:imageService:diaryService:notificationsService:webSocketsService:videoCallService:reachabilityService:sentryService:) at /Users/tikitu/work/minddistrict/code/ios-minddistrict/Abstractions/Global state management/Global application state.models.swift:175
#27	0x000000010fbd6cac in GlobalAppState.__allocating_init(network:storage:htmlFormatter:imageService:diaryService:notificationsService:webSocketsService:videoCallService:reachabilityService:sentryService:) ()
#28	0x000000010f55d6d5 in MDAppDelegate.setupDependencyInjection() at /Users/tikitu/work/minddistrict/code/ios-minddistrict/app/Features/Application startup/MDAppDelegate.swift:116
#29	0x000000010f55c1a5 in MDAppDelegate.application(_:didFinishLaunchingWithOptions:) at /Users/tikitu/work/minddistrict/code/ios-minddistrict/app/Features/Application startup/MDAppDelegate.swift:65
#30	0x000000010f55c553 in @objc MDAppDelegate.application(_:didFinishLaunchingWithOptions:) ()
#31	0x000000011b567731 in -[UIApplication _handleDelegateCallbacksWithOptions:isSuspended:restoreState:] ()
#32	0x000000011b5690cd in -[UIApplication _callInitializationDelegatesForMainScene:transitionContext:] ()
#33	0x000000011b56e4e6 in -[UIApplication _runWithMainScene:transitionContext:completion:] ()
#34	0x000000011adb2a4e in __111-[__UICanvasLifecycleMonitor_Compatability _scheduleFirstCommitForScene:transition:firstActivation:completion:]_block_invoke ()
#35	0x000000011adbb346 in +[_UICanvas _enqueuePostSettingUpdateTransactionBlock:] ()
#36	0x000000011adb2664 in -[__UICanvasLifecycleMonitor_Compatability _scheduleFirstCommitForScene:transition:firstActivation:completion:] ()
#37	0x000000011adb2fc0 in -[__UICanvasLifecycleMonitor_Compatability activateEventsOnly:withContext:completion:] ()
#38	0x000000011adb1332 in __82-[_UIApplicationCanvas _transitionLifecycleStateWithTransitionContext:completion:]_block_invoke ()
#39	0x000000011adb0fe9 in -[_UIApplicationCanvas _transitionLifecycleStateWithTransitionContext:completion:] ()
#40	0x000000011adb5d2e in __125-[_UICanvasLifecycleSettingsDiffAction performActionsForCanvas:withUpdatedScene:settingsDiff:fromSettings:transitionContext:]_block_invoke ()
#41	0x000000011adb6988 in _performActionsWithDelayForTransitionContext ()
#42	0x000000011adb5a95 in -[_UICanvasLifecycleSettingsDiffAction performActionsForCanvas:withUpdatedScene:settingsDiff:fromSettings:transitionContext:] ()
#43	0x000000011adbaa48 in -[_UICanvas scene:didUpdateWithDiff:transitionContext:completion:] ()
#44	0x000000011b56cdc8 in -[UIApplication workspace:didCreateScene:withTransitionContext:completion:] ()
#45	0x000000011b12402f in -[UIApplicationSceneClientAgent scene:didInitializeWithEvent:completion:] ()
#46	0x0000000116c11d25 in -[FBSSceneImpl _didCreateWithTransitionContext:completion:] ()
#47	0x0000000116c1bad6 in __56-[FBSWorkspace client:handleCreateScene:withCompletion:]_block_invoke_2 ()
#48	0x0000000116c1b300 in __40-[FBSWorkspace _performDelegateCallOut:]_block_invoke ()
#49	0x0000000112710db5 in _dispatch_client_callout ()
#50	0x00000001127142ba in _dispatch_block_invoke_direct ()
#51	0x0000000116c4d0da in __FBSSERIALQUEUE_IS_CALLING_OUT_TO_A_BLOCK__ ()
#52	0x0000000116c4cd92 in -[FBSSerialQueue _performNext] ()
#53	0x0000000116c4d327 in -[FBSSerialQueue _performNextFromRunLoopSource] ()
#54	0x000000010a30ddb1 in __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ ()
#55	0x000000010a30d633 in __CFRunLoopDoSources0 ()
#56	0x000000010a307cef in __CFRunLoopRun ()
#57	0x000000010a3074d2 in CFRunLoopRunSpecific ()
#58	0x0000000113c7c2fe in GSEventRunModal ()
#59	0x000000011b56ffc2 in UIApplicationMain ()
#60	0x00000001071cffdc in main at /Users/tikitu/work/minddistrict/code/ios-minddistrict/app/Features/Application startup/main.m:14
#61	0x000000011241d541 in start ()
#62	0x000000011241d541 in start ()```

**Comment by tikitu** [comment]
I've tried a full clean (clean build folder, remove Pods folder, remove all folders in derived data, re-run `pod install`, build app), with the 6.3.0 Kingfisher release: I still see the crash in the same place.

**Comment by onevcat** [comment]
Thanks for checking it. Umm, at last I cannot reproduce this problem at all.

However, since we already have several confirmation for the fix, version 6.3.1 was released to fix this.
